drivers/spi/spi_stm32.c

```diff
@@ -0,0 +1,584 @@
+/*
+ * Copyright (c) 2016 BayLibre, SAS
+ * Copyright (c) 2018 Bobby Noelte
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief STM32 SPI implementation
+ * @defgroup device_driver_spi_stm32 STM32 SPI inplementation
+ *
+ * A common driver for STM32 SPI. SoC specific adaptions are
+ * done by device tree and soc.h.
+ *
+ * @{
+ */
+
+/** Compile if at least one device is activated
+ * @code{.codegen}
+ * compatible = 'st,stm32-spi-fifo'
+ * codegen.if_device_tree_controller_compatible('SPI', compatible)
+ * @endcode{.codegen}
+ */
+/** @code{.codeins}@endcode */
```

Comment:
This is not OK.

In Zephyr it is the responsibility of the build system to determine which files should go into the build.

We can't have the build system believe we are using this source file. And then have some other system
that finds out we aren't going to use it after all and have it transformed into empty_file.c.

Instead, we need to find some way of communicating to the build system that this driver shall not be used.

To understand this better I need to understand under what situations this can occur. So I need to find some  board/sample/kconfig combination that can produce this situation.

Reply:
The idea is:

- SoC does have the corresponding (e.g. SPI) device -> CONFIG_SPI is set
- Application decides by DTS to activate none - this can only be seen in the device tree

So the easiest way is to check that at code generation time. The driver can not be build (regardless of inline code generation) as there are no defines defined for the driver´s device tree properties.

If you switch from KConfig to DTS you have to take the DTS info. KCONFIG will not even provide any control for this. You could scan the the DTS info the same way code generation does and provide CMake control.

Reply:
Thank you for trying to explain.

> SoC does have the corresponding (e.g. SPI) device -> CONFIG_SPI is set

When the SoC tries to communicate that it supports SPI through 'CONFIG_SPI', is it doing this through Kconfig?

> Application decides by DTS to activate none - this can only be seen in the device tree

Is the application doing this through application-specific DTS overlay's?

Also, is the above two points describing how Zephyr works today, or how it will work when the PR is merged?

Reply:
> When the SoC tries to communicate that it supports SPI through 'CONFIG_SPI', is it doing this through Kconfig?
> Is the application doing this through application-specific DTS overlay's?

Yes to all.

> Also, is the above two points describing how Zephyr works today, or how it will work when the PR is merged?

The PR does not change anything on the build process. This is IMHO how it works today.

The primary reason is that we have two build configuration sources (Kconfig, DTS) that have to be inline to allow the driver instantiation. But this discussion is not related to this PR and should perhaps be done in an extra issue aka. "Kconfig driver instantiation may clash with device tree".

- codegen.if_device_tree_controller_compatible('SPI', compatible) is one solution to the problem
- another solution would be a zephyr_sources(spi.c COMPAT "st,stm32-fifo") that looks into the compatibility properties of the actual device tree. Could be done by extract_dts_includes.py returning this information to CMake.

Reply:
"But this discussion is not related to this PR and should perhaps be done in an extra issue aka."

Agreed. There is something very wrong with the configuration system, but we can't block
users from using it until we have it working again.

Reply:
"Solution (2) is not really inline with the coding guidelines that require to use codegen before cmake."

Noted. But solution (2) is still the way to go.